Fix memory leak for RPCSEC_GSS

sachinpunadikar opened the pull request:
The GSS routines expects to release various fields been used by the
applicationi, after their use. Fixed the code to release such resources.

Signed-off-by: Sachin Punadikar <psachin@in.ibm.com>

dang commented on src/svc_auth_gss.c:
```diff
@@ -146,12 +146,13 @@ svcauth_gss_acquire_cred(void)
-				(void) gss_release_cred(&min_stat, &ancient_creds);
-			}
+		}
+		/* release the old creds anyway */
```
Why this?  If gss_acquire_cred() failed, then svcauth_gss_creds still points to old creds, so it will be double freed later, right?

sachinpunadikar replied:
It is not clear from the available documentation that on failure of gss_acquire_cred(), it will hold the previous value. So considering it may return NULL or any other garbage, its better to handle the cred free job outside of the success block.
As copying of old_creds & ancient_creds is done irrespective of success to the gss_acquire_cred() routine, its more logical to handle the free job outside of the success block.
I do not think there will be any double free. Only the call to gss_release_cred will fail with return code GSS_S_NO_CRED.
But if the creds are freed only on success of gss_acquire_cred(), then we end up leaking old_creds if svcauth_gss_creds do not point to the previous cred handle (But it is not cleraly documented either way)

sachinpunadikar replied:
Not changing this part in the updated patch

dang commented on src/svc_auth_gss.c:
```diff
@@ -251,6 +252,8 @@ svcauth_gss_accept_sec_context(struct svc_req *req,
 				"%s: display_name major=%u minor=%u", __func__,
 				maj_stat, min_stat);
 			gss_release_buffer(&min_stat, &gr->gr_token);
+			gss_delete_sec_context(&min_stat, &gd->ctx,
```
I don't think this is correct.  The context in gd is not created here, it's created elsewhere, and it stored in an SVCAUTH, which has it's own lifecycle.  The context will be destroyed when the svcauth is.

sachinpunadikar replied:
at line no 209, we are acquiring the context via gss_accept_sec_context. Only after success to that we are here to get the display name. Previous code was just returning from here on getting failure, without releasing the context. This is potential leak in failure case.

sachinpunadikar replied:
Looks like you are correct.

sachinpunadikar replied:
In any case no harm in deleting the context when we know it is not required. The code is least bothered about the return value of gss_delete_sec_context().

sachinpunadikar replied:
leaving the cleanup to be taken care by SVCAUTH related code.

dang commented on src/svc_auth_gss.c:
```diff
@@ -280,23 +283,29 @@ svcauth_gss_accept_sec_context(struct svc_req *req,
 		maj_stat =
-		    gss_sign(&min_stat, gd->ctx, GSS_C_QOP_DEFAULT, &seqbuf,
-			     &checksum);
+		    gss_get_mic(&min_stat, gd->ctx, GSS_C_QOP_DEFAULT, &seqbuf,
```
This change would be beyond the scope of a leak fix, I think.

sachinpunadikar replied:
I agree, I will push this as separate change.

dang commented on src/svc_auth_gss.c:
```diff
@@ -280,23 +283,29 @@ svcauth_gss_accept_sec_context(struct svc_req *req,
 
 		if (maj_stat != GSS_S_COMPLETE) {
 			gss_release_buffer(&min_stat, &gr->gr_token);
+			gss_delete_sec_context(&min_stat, &gd->ctx,
```
Again, this is part of an SVCAUTH and needs to be freed with it.

sachinpunadikar replied:
Updated patch rectified this

malahal commented on src/svc_auth_gss.c:
```diff
@@ -582,24 +598,53 @@ _svcauth_gss(struct svc_req *req, bool *no_dispatch)
+			if (gd_hashed)
+				(void)authgss_ctx_hash_del(gd);
+			unref_svc_rpc_gss_data(gd);
 			svcauth_gss_return(RPCSEC_GSS_CREDPROBLEM);
```
can we fix svcauth_gss_return() to do the clean up in a failure case?  Ideally, we should remove the macro and replace it with a goto block.

sachinpunadikar replied:
Yah, good suggestion. Will work on it.

dang commented on src/svc_auth_gss.c:
```diff
@@ -280,23 +283,29 @@ svcauth_gss_accept_sec_context(struct svc_req *req,
 		if (checksum.length > MAX_AUTH_BYTES){
 			gss_release_buffer(&min_stat, &gr->gr_token);
+			gss_release_buffer(&min_stat, &checksum);
+			gss_delete_sec_context(&min_stat, &gd->ctx,
```
And here.

Commit:
Use "gss_get_mic" instead of "gss_sign"

"gss_sign" is a deprecated GSS-API function. This needs to be replaced by
"gss_get_mic".

Signed-off-by: Sachin Punadikar <psachin@in.ibm.com>

Commit:
Fix memory leak for RPCSEC_GSS

The GSS routines expects to release various fields been used by the
applicationi, after their use. Fixed the code to release such resources.

Signed-off-by: Sachin Punadikar <psachin@in.ibm.com>

Commit:
Do not use macro "svcauth_gss_return"

The svcauth_gss_return macro was used to return from function _svcauth_gss.
But this macro do not carry out required cleanup in failure cases.
Removed this macro and used goto block, which takes care of required cleanup.

Signed-off-by: Sachin Punadikar <psachin@in.ibm.com>

sachinpunadikar commented:
Please review the 3 patches

dang commented on src/svc_auth_gss.c:
```diff
@@ -406,13 +397,16 @@ _svcauth_gss(struct svc_req *req, bool *no_dispatch)
 	OM_uint32 min_stat;
 	bool gd_locked = false;
```
You're not using gd_locked anymore.  You can remove it.

sachinpunadikar replied:
Good suggestion. Removed it

dang commented on src/svc_auth_gss.c:
```diff
@@ -628,15 +653,22 @@ _svcauth_gss(struct svc_req *req, bool *no_dispatch)
 	}
-
-	svcauth_gss_return(AUTH_OK);
+gd_free:
```
Can we rename one of these?  Having them be one letter different makes them hard to track, and it's going to result in someone using the wrong one.

sachinpunadikar replied:
Renamed gc_free to cred_free label.

dang commented on src/svc_auth_gss.c:
```diff
@@ -628,15 +653,22 @@ _svcauth_gss(struct svc_req *req, bool *no_dispatch)
 		 */
 		unref_svc_rpc_gss_data(gd);
```
You should set gd_hashed to false here, in case code is added after this in the function.

sachinpunadikar replied:
Done.
